### DISCUSSION TITLE
Migration to v9: cannot figure out how to use "files" and "ignores" options

### DISCUSSION THREAD

**Discussion by kleinfreund** [eslint/eslint#19028] [Q&A] [answered]
I'm migrating a project from ESLint 8 to 9 and I'm not getting anywhere with replicating my current ESLint 8 configuration.

The project is written in TypeScript using ES modules and is linted with `eslint --ext .ts .` while `dist/` and `coverage/` are ignored via .eslintignore.

According to https://eslint.org/docs/latest/use/configure/migration-guide#--ext, the `--ext` CLI argument becomes a matter of file-based configuration. Because I was completely stuck getting my existing config to work, I threw out _everything_ and started over with a bare-bones configuration without plugins or TypeScript integration.

```js
export default [
	{
		files: ['src/'],
		rules: {
			'comma-dangle': ['error', 'always-multiline'],
		},
	}
]
```

I ran `npx eslint`.

This doesn't lint the files in src/ and also lints files outside of it (specifically, the files in coverage/).

Next, I tried specifically targeting TS files:

```js
export default [
	{
		files: ['src/**/*.ts'],
		rules: {
			'comma-dangle': ['error', 'always-multiline'],
		},
	}
]
```

Now, .ts files are linted (if I use a directory (i.e. `src/`) to scope files I expect all files inside of it to be linted, not just JavaScript files).

Meanwhile, explicitly ignoring `coverage/` doesn't work either:

```js
export default [
	{
		files: ['src/**/*.ts'],
		ignores: ['coverage/'], // This time, being explicit and using `'coverage/**/*.js'` doesn't change the outcome.
		rules: {
			'comma-dangle': ['error', 'always-multiline'],
		},
	},
]
```

So, how do you _actually_ select and ignore files?

The documentation at https://eslint.org/docs/latest/use/configure/migration-guide#ignoring-files shows how ignoring _should_ work but I cannot figure out how to apply it. _Except for file paths that are inside `files` (i.e. files I actually don't care to ignore)_, nothing I put in `ignores` seems to apply.

I can only manage to get coverage/ to be ignored by passing `--ignore-pattern coverage/` to the CLI.

Conversely, why is there some secret magic filtering going on when I used `src/` (where TS files are ignored) compared to `src/**/*.ts`? That's very confusing.

**Comment by kleinfreund** [comment, marked as answer]
One answer I immediately found after opening this discussion was https://github.com/eslint/eslint/discussions/18456#discussioncomment-10799656: it looks like the `ignores` option must appear in a separate object like so:

```js
export default [
	{
		ignores: ['coverage/', 'dist/'],
	},
	{
		files: ['src/**/*.ts'],
		rules: {
			'comma-dangle': ['error', 'always-multiline'],
		},
	},
]
```

Now what I still don't understand is why `src/` doesn't work but `src/**/*.ts` does.

**Reply by mdjermanovic** [reply]
> Now what I still don't understand is why `src/` doesn't work but `src/**/*.ts` does.

Patterns in `files` match only files, so `src/` has no effect.

**Reply by kleinfreund** [reply]
And is it also the case that I have to be explicit about the extension? The following also doesn't lint .ts files when I expected this to lint virtually all files within `src/`.

```js
export default [
	{
		ignores: ['coverage/', 'dist/'],
	},
	{
		files: ['src/**/*'],
		rules: {
			'comma-dangle': ['error', 'always-multiline'],
		},
	},
]
```

**Reply by mdjermanovic** [reply]
> And is it also the case that I have to be explicit about the extension?

Yes. Precisely, in order to be included in linting, files other than .js, .mjs, and .cjs (which are included by default) need to match at least one config object, and in it at least one `files` pattern that isn't `*` and doesn't end with `/*` or `/**`.

https://eslint.org/docs/latest/use/configure/configuration-files#specifying-files-with-arbitrary-extensions

In your example, `src/**/*` does match .ts files, but doesn't include them in linting because it ends with `/*`.

**Comment by mdjermanovic** [comment]
> The documentation at https://eslint.org/docs/latest/use/configure/migration-guide#ignoring-files shows how ignoring _should_ work but I cannot figure out how to apply it

Thanks for the input! I think https://github.com/eslint/eslint/pull/19032 will help at this point.
